### ISSUE TITLE
Allowing sprites in StringTable

### ISSUE THREAD

**Issue by Brett208** [OutpostUniverse/OPHD#637]
I was considering allowing a single sprite in StringTable instead of just text. This would allow using the ore symbols in the Smelter StructureInspector instead of just the text.

**Comment by DanRStevens** [comment]
I was thinking about the recently, I suspected this sort of issue might come up soon.

I think what we really want, is a grid layout control. The cells would be containers for arbitrary UI elements. With the current implementation, it's like a grid layout where all cells are filled specifically with `Label` controls.

This relates to the idea of developing more general layout containers.

**Comment by Brett208** [comment]
I would be ok abstracting a grid layout. Perhaps it replaces `StringTable` altogether or perhaps there is a child of Grid that simplifies setup for tables only containing strings.

Image doesn't inherit from control though. So to be a generic control grid that could handle images, it would require adding an image control.

Another option would be to add a ScreenObject class that represents the following:
 * Position on screen
 * rectangle size
 * virtual draw function
 * virtual update function

**Comment by DanRStevens** [comment]
Yes, I've been thinking about how screen position and size should relate to the control objects. In regards to object size, some objects have a natural size, such as images and labels (updated when text changes), while many do not. Some also have a minimum size. That got me thinking maybe size should be part of the object itself, so the object can have more control over it. Meanwhile position seems largely controlled by the container. Not sure I would want to split size and position between different objects though. Plus, if the container holds the object position, it becomes harder to determine position given a control. I'm leaning towards size and position being members of the control itself, and high up in the `Control` hierarchy.

In terms of the `update` and `draw` functions, I think the `draw` function should be marked as `const`.

I think you're right about needing an `Control` form of `Image`. I've noticed the `Sprite` class as implemented is more like a `Control` than a resource. We should probably split that object into two. Generic resources can be thought of as cached common data, without associated screen position (though may have a natural concept of size). A `Control` may reference such resources, and will have associated position and state data.

Another thing I noted, is that we've added a `Label` control, but we don't really use it. A lot of text is manually positioned, and that position is calculated when it is drawn. If we split `update` and `draw`, then `Label` positioning would probably be associated more with `update`, or more likely actual changes to `Label` text data. The `draw` function would then be using pre-calculated coordinates. I suppose technically that would inflate the sizes of objects though, since they would necessarily contain pre-calculate position/size state data, rather than calculating on the fly. Or at least that's true if the objects themselves contain the position and size, rather than the container (which might calculate things on the fly).

Related, the base `Control` class has methods for text fields, but not every `Control` contains text. Maybe the text related methods need to be moved elsewhere. Not sure if we should have a generic `TextControl` intermediate or not. Possibly not. Also not sure how much we should break down subclasses. For the `Button` control, should we have separate classes for text buttons versus image buttons? The current design combines them, with an optional `Image`. If we start splitting things too much, we may end up with multiple inheritance issues, where controls have base classes for different orthogonal concerns (text control, image control, text and image control etc.).

The ideas still feel pretty incoherent so far. There doesn't seem to be a clear and obvious best design.

**Comment by Brett208** [comment]
I was working on this last week and designed a prototype ImageControl, which I didn't get far enough in to actually test outside of compiling.

I was realizing that Sprite is for an animated sprite and Image is for an unanimated Sprite. I was hoping to see Image and Sprite be organized a little closer than they are in calling convetions.

To convert StringTable into a more generic UIGrid, it would be helpful to provide an abstract TextControl. This would provide unified access to setting the font, font color, and maybe a few other text related activities. This would facilitate applying StringTable::SetColumnFont and other functions to all controls in a range that are primarily designed to show text.

**Comment by DanRStevens** [comment]
I agree with your assessments. These are things I'd like to work on, and could probably use a bit of help and feedback with.

----

The current `Sprite` class design mixes elements of both `Control` and resource. The `Image` class doesn't have a position on screen, and so is unlike a `Control`, while `Sprite` does have a position. On the other hand, the generic animation info is loaded into the `Sprite` class, much like generic `Image` info is loaded into the `Image` class. This leads to a lot of duplicated data in similar `Sprite` instances. I was thinking these should be split, so generic animation information exists in a resource class, while a `Control` like class can reference that data and supply an associated on-screen position.

In the [nas2d-tests](https://github.com/lairworks/nas2d-tests) repository, I had noticed the zombie shooter sample was parsing the zombie sprite data 5 times, once for each zombie shown on screen. I think it may have been updated to not reparse the XML data multiple times, though the class structure still needs update.

I feel like we may need to rework some of the terminology. I think of a sprite as an image with transparency. An animated sprite seems like a further specialization. I'm not sure an animated image which is not a sprite would be very common, so perhaps we can assume that any animated objects will refer to sprites. Of course, the `Image` object has support for alpha channels, so there doesn't seem to be a need to differentiate between images and images with transparency at the class level.

In terms of interface, I would like to be able to `draw` objects, passing a `Renderer` reference, and I would like that to be possible on `const` objects. Whether such a `draw` method goes on the resource objects themselves (in which case, a position would also need to be passed), or if the method goes on a `Control` version of the resource, I haven't really decided. Maybe have one method delegate to the other. I believe change to object state should be handled separately by an `update` method, which would not be `const`. For an `Image` class, there isn't anything to update. For a `Sprite` class (or whatever `Control` form of it might exist after splitting or renaming), such an `update` method could be used to iterate through the frames.

----

I was also thinking we should have some kind of `TextControl` base class. In particular, there are some text methods in `Control`, which doesn't seem like an appropriate place.

One concern I have here, is a layout class might work with arbitrary `Control` objects, while the text based methods would only apply to `TextControl` objects. I suppose we could use `dynamic_cast` to check for `Control` classes of the correct type, and only set text or `Font` properties if they apply. Normally I like to avoid `dynamic_cast`, though I'm not sure how to avoid it in this case, and still maintain a sane API. On the plus side though, people should only be calling methods that might use `dynamic_cast` in cases where `dynamic_cast` would actually apply and be useful. If you have a grid layout of images, with no text based controls, you shouldn't be calling methods that set text, text color, or font.

**Comment by Brett208** [comment]
Looks like this issue will wonder quite a bit off topic. I think there is a lot more pre-work to completing this than I thought...

I would consider 2D graphics drawn to screen as sprites. Maybe rename `Image` to `Sprite`. The current `Sprite` class could be renamed `AnimatedSprite`. That would make the terminology more consistent. 

If there is a desire to have a resource behind a Sprite that doesn't understand screen coordinates, I would consider maybe `Texture` or `Texture2D` if possibility exists for 3 dimensional support in future. A sprite would have 1 texture at a time and an AnimatedSprite could contain multiple textures.

Assuming above naming convention, I like the idea of a `void Sprite::draw(Renderer& renderer) const` and `void AnimatedSprite::draw(Renderer& renderer` const. 

I'm not sure how to define the control class. I suspect there is room to have a basic `Sprite` class that allows only drawing to screen and then a `SpriteControl` that adds control functionality on top of it. I think the `Control` interface maybe should add concepts like in focus, hovered over, etc? I haven't really fleshed out what Control should be.

In C# it seems more acceptable to use dynamic_cast as a discriminator to only select items within a certain class hierarchy (I don't know that it is a one for one comparison to dynamic_cast). I've used it with others in C# projects who thought it was normal practice. I have no issue with using dynamic_cast in this method. I like the idea of preserving some of the text helper functions understanding they would not apply to all cells based on grid contents. Although it also wouldn't be wrong to just remove those fields either, but would require a lot more manual entries to affect rows or columns of text.

**Comment by DanRStevens** [comment]
I kind of like the idea of naming things `Sprite` and `AnimatedSprite`. It's reasonably clear and accurate. My one concern is that people might naturally go looking for an `Image` class, so not having one might be confusing.

We are currently using the name "texture" internally for the associated OpenGL handle. From what a gather of how the API works, `Image` is basically the raw data in main memory, while "texture" is when it's loaded into graphics memory, which may convert it to a hardware specific optimized format.

It terms of the Resource/`Control` difference, the main aspect is position. You can have multiple controls all referencing the same resource, at different positions on screen. The actual image data itself doesn't need to be duplicated, it just gets drawn at different positions, using the coordinates of each corresponding `Control`.

A Resource may have a natural concept of `size`, but it lacks `position`. I suppose it's a little like the difference between a class and an object.

In terms of `Control` attributes, I think the most basic are position and size. Things like text features only apply to certain controls. Similarly the ability to accept focus only applies to certain controls. Take the `Label` control for instance, which doesn't accept focus. Mind you, maybe a "Control" is something that accepts focus, while merely having a position and size is for some yet unnamed base class of `Control`.

I like that we're thinking about focus though, because that's an issue that's crossed my mind from time to time. The current controls seem a bit ad hoc about focus.

I think you're right about these ideas needing a bit of pre-work to handle them smoothly. I've tried tinkering with a few ideas before, though they often started to cascade into rather large change sets, which I generally backed out of.

**Comment by ldicker83** [comment]
When we say 'sprite' are we talking about the actual NAS2D::Sprite class or are we speaking of just images or image slices?

**Comment by DanRStevens** [comment]
I think the discussion is more conceptual, and extends to arbitrary UI elements, not just `NAS2D::Sprite`. I think we should be able to put arbitrary `Control` instances in the table cells.

Yes, there was some side discussion about the design of `NAS2D::Sprite`, which kind of relates, though is a bit off topic.

----

In terms of `NAS2D::Sprite`, there was discussion about splitting the static shared resource portion of it from the more control specific position and animation state portion. Think how the Zombie sample game used to parse the sprite XML file 5 times, once per each zombie displayed in the game, even though the resources were identical and could be shared between them. There is now some splitting internally along those lines, though it isn't exposed well. Basically, the design of `NAS2D::Sprite` is unlike the design of `NAS2D::Image`, and the discussion was about splitting the control to resolve those differences.

I don't plan to split between animated and non-animated sprites. We already have `Image` versus `Sprite` that basically handles that. I think we were looking more at potentially renaming things if it might add clarity. The name `Sprite` implies transparency, though doesn't imply animation, so we were looking at names that might be more descriptive. In addition, if we split the `NAS2D::Sprite` class so the resource data is stored separately form the control data, then we'd have an extra class to name. Similar to if we added an image control, which currently doesn't exist (only `NAS2D::Image` resources, which have no associated screen position).

**Comment by ldicker83** [comment]
What's confusing me is why Sprite is being considered or thought of as a UI Control? It's not at all part of that concept. Sprite is defined as an image that's not part of a background and may or may not be animated. That's effectively what the sprite is so treating it as a Control makes no sense to me.

**Comment by Brett208** [comment]
In game libraries, I'm used to seeing a Sprite class that is a static image with a screen position and ability to render itself. Which is why I was thinking maybe the NAS2D::Sprite could be renamed to AnimatedSprite. That way Sprite could be repurposed to be a static image with associated position and ability to draw itself. 

Sometimes, it would be nice to have a sprite control so you can treat a sprite as a layout object within a control grid. It would have more features than just drawing to screen.

**Comment by DanRStevens** [comment]
`Sprite` doesn't derive from `Control`, but it contains state similar to `Control` derived classes, and is responsible for drawing itself using an `update` method. Effectively it could be made to derive from `Control` and would naturally fit in. (Not a suggestion, just an observation).

Looking back at it now, I see `Sprite` does not actually contain a `position` (it's a parameter to `update`), but it does contain other instance specific state variables such as:
- `mCurrentAction`
- `mCurrentFrame`
- `mTimer`
- `mAnimationCompleteCallback`
- `mColor` (this could be considered either a specific or generic field)
- `mRotationAngle`

In contrast, there are generic fields for the static resource data:
- `mSpriteName`
- `mActions` (and the associated frame lists)

The generic data in `mActions` is the proper "Resource" data. This data needs to be stored separately from the instance specific data above to avoid duplication and wasted space. Currently that's done with an internal private cache for the `mActions` data. It's mainly the `mActions` data structure that should form the "Resource" class, perhaps using a name such as "Animation". The other instance specific fields could be split into a new class, perhaps retaining the "Sprite" name or maybe "AnimatedSprite". The sprite class would contain a const reference to the static animation data. That is now currently true, in that `mActions` is a pointer to const data, though it's done using private internal caches, which is a bit of a rigid design. There is no way to construct animations from alternate data sources, such as building one up in memory. The only access to the private internal cache is through the file loading mechanism. By splitting the current `Sprite` class into those two components, we'd gain more control in how animations could be loaded or created, and how they are cached.

If the class was split, (and this is an observation, not a suggestion), the instance specific portion could potentially be moved to OPHD and made to inherit from `Control`, and would basically just work. The one change might be to add position as a member variables, which might as well be done since `mCurrentFrame` and `mRotationAngle` are already pretty instance specific, much like position would be. What I'm saying is this portion of the existing `Sprite` class is very similar to how `Control` derived classes work.

In contrast to `NAS2D::Sprite`, the `NAS2D::Image` class has no instance specific data. It is a pure "Resource" class.

You could of course build an "image" `Control` as well, which might be done by combining a const reference to a `NAS2D::Image` along with instance specific data, such as position, and maybe rotation angle. Such a new class would be used very similar to how `Sprite` is currently used now.
